File: src/main/java/no/nav/syfo/rest/feil/ThrowableMapper.java
package no.nav.syfo.rest.feil;

import org.slf4j.Logger;

import javax.ws.rs.WebApplicationException;
import javax.ws.rs.core.Response;
import javax.ws.rs.ext.ExceptionMapper;
import javax.ws.rs.ext.Provider;

import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
import static javax.ws.rs.core.Response.status;
import static no.nav.syfo.rest.feil.Feilmelding.Feil.GENERELL_FEIL;
import static no.nav.syfo.rest.feil.Feilmelding.NO_BIGIP_5XX_REDIRECT;
import static org.slf4j.LoggerFactory.getLogger;

@Provider
public class ThrowableMapper implements ExceptionMapper<Throwable> {

    private static final Logger LOG = getLogger(ThrowableMapper.class);

    @Override
    public Response toResponse(Throwable e) {
        LOG.error("Uventet feil", e);

        return status(statuskode(e))
                .type(APPLICATION_JSON)
                .entity(new Feilmelding().withFeil(GENERELL_FEIL))
                .header(NO_BIGIP_5XX_REDIRECT, true)
                .build();
    }

    private int statuskode(Throwable e) {
        if (e instanceof WebApplicationException) {
            return ((WebApplicationException) e).getResponse().getStatus();
        } else {
            return GENERELL_FEIL.status.getStatusCode();
        }
    }
}

File: src/main/kotlin/no/nav/syfo/api/exception/ThrowableMapper.kt
package no.nav.syfo.api.exception

import no.nav.syfo.api.exception.Feilmelding.Feil
import org.slf4j.LoggerFactory
import javax.ws.rs.WebApplicationException
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response
import javax.ws.rs.ext.ExceptionMapper
import javax.ws.rs.ext.Provider

@Provider
class ThrowableMapper : ExceptionMapper<Throwable> {
    override fun toResponse(e: Throwable): Response {
        LOG.error("Uventet feil", e)
        return Response.status(statuskode(e))
            .type(MediaType.APPLICATION_JSON)
            .entity(Feilmelding().withFeil(Feil.GENERELL_FEIL))
            .header(Feilmelding.NO_BIGIP_5XX_REDIRECT, true)
            .build()
    }

    private fun statuskode(e: Throwable): Int {
        return if (e is WebApplicationException) {
            e.response.status
        } else {
            Feil.GENERELL_FEIL.status.statusCode
        }
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(ThrowableMapper::class.java)
    }
}
